tslib/react/src/utils/trialFilter.ts

```diff
@@ -262,3 +293,103 @@ export const useObjectiveAndUserAttrTargetsFromStudies = (
   )
   return [targetList, selectedTarget, setTargetIdent]
 }
+
+const getIsDominatedND = (normalizedValues: number[][]) => {
+  // Fallback for straight-forward pareto front algorithm (O(N^2) complexity).
+  const isDominated: boolean[] = []
+  for (const values0 of normalizedValues) {
+    const dominated = normalizedValues.some((values1) => {
+      if (values0.every((value0, k) => values1[k] === value0)) {
+        return false
+      }
+      return values0.every((value0, k) => values1[k] <= value0)
+    })
+    isDominated.push(dominated)
+  }
+  return isDominated
+}
+
+const getIsDominated2D = (normalizedValues: number[][]) => {
+  // Fast pareto front algorithm (O(N log N) complexity).
+  const sorted = normalizedValues
+    .map((values, i) => [values[0], values[1], i])
+    .sort((a, b) =>
+      a[0] > b[0]
+        ? 1
+        : a[0] < b[0]
+          ? -1
+          : a[1] > b[1]
+            ? 1
+            : a[1] < b[1]
+              ? -1
+              : 0
+    )
+  let maxValueSeen0 = sorted[0][0]
+  let minValueSeen1 = sorted[0][1]
+
+  const isDominated: boolean[] = new Array(normalizedValues.length).fill(false)
+  for (const values of sorted) {
+    if (
+      values[1] > minValueSeen1 ||
+      (values[1] === minValueSeen1 && values[0] > maxValueSeen0)
+    ) {
+      isDominated[values[2]] = true
+    } else {
+      minValueSeen1 = values[1]
+    }
+    maxValueSeen0 = values[0]
+  }
+  return isDominated
+}
+
+const getIsDominated1D = (normalizedValues: number[][]) => {
+  const best_value = Math.min(...normalizedValues.map((values) => values[0]))
+  return normalizedValues.map((values) => values[0] !== best_value)
+}
+
+const getIsDominated = (normalizedValues: number[][]) => {
+  if (normalizedValues.length === 0) {
+    return []
+  }
+  if (normalizedValues[0].length === 1) {
+    return getIsDominated1D(normalizedValues)
+  }
+  if (normalizedValues[0].length === 2) {
+    return getIsDominated2D(normalizedValues)
+  }
+  return getIsDominatedND(normalizedValues)
+}
+
+function getIsDominatedTrials(study: Optuna.Study): boolean[] {
+  return (() => {
+    const trials: Optuna.Trial[] = study ? study.trials : []
+    const feasibleTrials: Optuna.Trial[] = getFeasibleTrials(trials, study)
+
+    const normalizedValues: number[][] = []
+    for (const t of feasibleTrials) {
+      if (t.values && t.values.length === study.directions.length) {
+        const trialValues = t.values.map((v, i) => {
+          return study.directions[i] === "minimize"
+            ? (v as number)
+            : (-v as number)
+        })
+        normalizedValues.push(trialValues)
+      }
+    }
+
+    return getIsDominated(normalizedValues)
+  })()
+}
```

Comment:
With the current optuna-dashboard implementation, the pareto front computation is in optuna-dashboard and cannot be referenced from tslib, so I added the same code to tslib.
If there is no problem, is it ok to change the pareto front calculation part of the pareto front plot in optuna-dashboard to the one in tslib as well?